### PR TITLE
sbt-based PR validation job

### PR DESCRIPTION
**Pull request by szeiger**
Testing the sbt-based PR validation...

### CODE REVIEW

**Comment by szeiger** [comment]
@SethTisue I think if you add the publishing stuff to the script that we figured out on Friday, it should actually work. Can you try that?

Remaining issues: Enable optimization & do not publish docs

**Comment by SethTisue** [comment]
commit added, let's see what happens

**Comment by SethTisue** [comment]
looks like `publish-core` went fine, but then of course `validate-test` couldn't run because we aren't yet publishing to https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/

**Comment by SethTisue** [comment]
getting closer:

```
[error] Unable to find credentials for [Artifactory Realm @ scala-ci.typesafe.com].
```

**Comment by SethTisue** [comment]
yay a bunch of JARs actually got published to the scala-ci artifactory :tada:

but then it [bombed](https://scala-ci.typesafe.com/job/scala-2.11.x-validate-publish-core/2241/console) with

```
java.lang.RuntimeException: Repository for publishing is not specified.
    at scala.sys.package$.error(package.scala:27)
    at sbt.Classpaths$$anonfun$getPublishTo$1.apply(Defaults.scala:1493)
    at sbt.Classpaths$$anonfun$getPublishTo$1.apply(Defaults.scala:1493)
    at scala.Option.getOrElse(Option.scala:120)
    at sbt.Classpaths$.getPublishTo(Defaults.scala:1493)
    at sbt.Classpaths$$anonfun$62.apply(Defaults.scala:1165)
    at sbt.Classpaths$$anonfun$62.apply(Defaults.scala:1165)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:439)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:662)
[error] (root/*:publishConfiguration) Repository for publishing is not specified.
```

EDIT: this is even though the root project has `publishArtifact := false`

**Comment by SethTisue** [comment]
hmm so artifacts published during https://scala-ci.typesafe.com/job/scala-2.11.x-validate-publish-core/2244/console showed up at e.g. https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/org/scala-lang/scala-library/2.11.8-f0d538a-SNAPSHOT/ but then https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/1609/console didn't find all (any?) of the artifacts it wanted

**Comment by SethTisue** [comment]
perhaps the problem is the mix of `-1` and `-2` suffixes (here's [ant not doing that](https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/org/scala-lang/scala-compiler/2.11.8-3a2fccc-SNAPSHOT/)). but what is causing those suffixes?

**Comment by SethTisue** [comment]
ah, this appears to be https://github.com/sbt/sbt/issues/2088

**Comment by adriaanm** [comment]
can we just drop the -SNAPSHOT?

**Comment by szeiger** [comment]
Dropping -SNAPSHOT would mean that we'd have to change the existing infrastructure that relies on these version numbers. It's the first time I've seen these "-1" and "-2" suffixes. I have no idea what they are and where they come from. I'll have to check the sbt sources.

File: build.sbt [hunk 1]
```diff
@@ -81,6 +81,8 @@ val scalacheckDep = withoutScalaLang("org.scalacheck" %% "scalacheck" % versionN
   * real publishing should be done with sbt's standard `publish` task. */
 lazy val publishDists = taskKey[Unit]("Publish to ./dists/maven-sbt.")
 
+lazy val publishDestination = settingKey[Option[Resolver]]("determines publishTo in multiple places")
```
Comment (szeiger) [inline build.sbt]: @SethTisue  What is the reason for using this over publishTo? It is only referenced in one place (for setting publishTo)

**Comment by szeiger** [comment]
As far as I can tell this looks good. The integrate-ide job (and consequently validate-main) has been failing for several weeks. @SethTisue WDYT? Shall I clean up the code and squash?

(Still looking into the optimization and docs stuff)

**Comment by szeiger** [comment]
Docs disabled, optimization enabled. The doc artifacts are gone and validate-test passes, so this should be ready for squashing and merging.

**Comment by adriaanm** [comment]
:cake:!

**Comment by szeiger** [comment]
Cleaned up and squashed.

**Comment by SethTisue** [comment]
/nothingtoseehere

(integrate-ide failure is unrelated to the PR)

**Comment by dwijnand** [comment]
What was done to avoid sbt/sbt#2088?

**Comment by szeiger** [comment]
@dwijnand Setting a consistent build.timestamp: https://github.com/szeiger/scala/blob/wip/sbt-pr-validation/project/ScriptCommands.scala#L14

**Comment by SethTisue** [comment]
@szeiger may I suggest some light testing with sbt 0.13.10-RC1? it'd be a bit awkward if we merged this and then were immediately stuck on 0.13.9 because of some regression

**Comment by szeiger** [comment]
@SethTisue No luck with 0.13.10-RC1: https://gist.github.com/szeiger/d494e70da91eb1ec9d06

**Comment by dwijnand** [comment]
You want to try 0.13.10-M1 (but @eed3si9n might want to cut a new release).

Or you might want to wait for a successful community build for 0.13.10, see https://github.com/scala/community-builds/pull/210.

**Comment by szeiger** [comment]
I was able to build it successfully with sbt 0.13.10-M1 but not with 0.13.10-RC1. /cc @eed3si9n

**Comment by dwijnand** [comment]
That's good news. Thank you @szeiger.

**Comment by adriaanm** [comment]
time to merge 2.11.x into 2.12.x?

**Comment by dwijnand** [comment]
Cool!

**Comment by SethTisue** [comment]
> time to merge 2.11.x into 2.12.x?

we're not that close yet. this PR makes validate-publish-core sbt-based, but doesn't update validate-test (not to mention integrate-bootstrap and integrate-windows). we can discuss overall status at https://github.com/scala/scala-dev/issues/45

**Pull request merged**